test/e2e/kb/version_upgrade_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kb
+
+import (
+	"testing"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/kibana/label"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/kibana"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestVersionUpgrade(t *testing.T) {
+	name := "test-version-upgrade"
+	esBuilder := elasticsearch.NewBuilder(name).
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithVersion("7.4.1")
+
+	kbBuilder := kibana.NewBuilder(name).
+		WithElasticsearchRef(esBuilder.Ref()).
+		WithNodeCount(3).
+		WithVersion("7.4.0")
+
+	ns := client.InNamespace(kbBuilder.Kibana.Namespace)
+	matchLabels := client.MatchingLabels(map[string]string{
+		common.TypeLabelName:      label.Type,
+		label.KibanaNameLabelName: kbBuilder.Kibana.Name,
+	})
+
+	var observations []int
+	w := test.NewWatcher("watch pods readiness", 1*time.Second,
+		func(k *test.K8sClient, t *testing.T) {
+			if pods, err := k.GetPods(ns, matchLabels); err != nil {
+				t.Logf("got error: %v", err)
+			} else {
+				ready := 0
+				for _, pod := range pods {
+					if k8s.IsPodReady(pod) {
+						ready++
+					}
+				}
+				observations = append(observations, ready)
+			}
+		},
+		func(k *test.K8sClient, t *testing.T) {
+			assert.Contains(t, observations, 0)
```

Comment:
I'm not sure I understand what this is doing. From what I read we check that, at least once, we observed 0 pods ready? I guess this would mean "we expect downtime". What if the watcher does not catch the downtime period? I know it's running every 1 sec but stil, sounds a bit arbitrary.

Another approach would be to store a list of all current Pods versions in one observation. Then at the end go through all observations and make sure there is no observation that contains 2 different versions. Meaning 2 different versions of Kibana where running at the same time. This is catching "never 2 different Kibana versions simultaneously" instead of "expect a downtime", which is closer to what we want, I think.

Reply:
I think both approaches are equivalent for all practical purposes, so I'll change it, but for the record, seeing downtime positively confirms there were no two same versions running. Not seeing two versions simultaneously confirms just that, that we didn't observe two versions running at the same time (but it's not implicating that it didn't happen).

Reply:
Chatted offline, we will have both.